website/notifications/utils.py

```diff
@@ -0,0 +1,242 @@
+from website.notifications.constants import SUBSCRIPTIONS_AVAILABLE, NOTIFICATION_TYPES, USER_SUBSCRIPTIONS_AVAILABLE
+from modularodm import Q
+from modularodm.exceptions import NoResultsFound
+
+
+class NotificationsDict(dict):
+    def __init__(self):
+        super(NotificationsDict, self).__init__()
+        self.update(messages=[], children=collections.defaultdict(NotificationsDict))
+
+    def add_message(self, keys, messages):
+        """
+        :param keys: ordered list of project ids from parent to node (e.g. ['parent._id', 'node._id'])
+        :param messages: built email message for an event that occurred on the node
+        :return: nested dict with project/component ids as the keys with the message at the appropriate level
+        """
+        d_to_use = self
+        for key in keys:
+            d_to_use = d_to_use['children'][key]
+        if not isinstance(messages, list):
+            messages = [messages]
+        d_to_use['messages'].extend(messages)
+        return True
+
+
+def to_subscription_key(uid, event):
+    return str(uid + '_' + event)
+
+
+def from_subscription_key(key):
+    parsed_key = key.split("_", 1)
+    return {
+        'uid': parsed_key[0],
+        'event': parsed_key[1]
+    }
+
+
+@contributor_removed.connect
+def remove_contributor_from_subscriptions(contributor, node):
+    """ Remove contributor from node subscriptions unless the user is an
+        admin on any of node's parent projects.
+    """
+    if not contributor._id in node.admin_contributor_ids:
+        node_subscriptions = get_all_node_subscriptions(contributor, node)
+        for subscription in node_subscriptions:
+            subscription.remove_user_from_subscription(contributor)
+
+
+@node_deleted.connect
+def remove_subscription(node):
+    Subscription.remove(Q('object_id', 'eq', node._id))
+
+
+def get_configured_projects(user):
+    """ Filter all user subscriptions for ones that are on parent projects and return the project ids
+    :param user: modular odm User object
+    :return: list of project ids for projects with no parent
+    """
+    configured_project_ids = []
+    user_subscriptions = get_all_user_subscriptions(user)
+    for subscription in user_subscriptions:
+        node = Node.load(subscription.object_id)
+        if node and not node.is_deleted: # if node is deleted, the subscription should be removed anyway
+            parent = node.parent_node
+            has_child_node_subscriptions = node.has_child_node_subscriptions != []
+
+            # Include private parent ids so user subscriptions on the node are still displayed
+            if user not in subscription.none and parent and not parent.parent_node and not parent.has_permission(user, 'read') and parent._id not in configured_project_ids:
+                configured_project_ids.append(parent._id)
+
+            elif not parent and subscription.object_id not in configured_project_ids and has_child_node_subscriptions:
```

Comment:
@samanehsan Is this check correct? It is causing this test to fail: https://github.com/CenterForOpenScience/osf.io/blob/7ae9a942f62fe5373f1df17f2c69ed1f28088b79/tests/test_notifications.py#L451 . Specifically, I think the check for `has_child_components` is causing parent projects to not be included in the return value.